eth/beacon/state_machines/forks/serenity/epoch_transitions.py

```diff
@@ -0,0 +1,224 @@
+from eth.beacon.types.states import BeaconState
+from eth.beacon.state_machines.configs import BeaconConfig
+from eth.beacon.helpers import (
+    get_effective_balance,
+    get_active_validator_indices,
+    get_attestation_participants,
+    get_block_root,
+    total_balance,
+)
+from eth.beacon.constants import TWO_POWER_64
+
+
+def get_epoch_boundary_attesting_balances(
+        state: BeaconState,
+        config: BeaconConfig) -> Tuple[Gwei, Gwei]:
+    """
+    previous_epoch_boundary_attesting_balance:
+        SELECT sum(v.balance)
+        FROM Validator as v
+        RIGHT JOIN Attestations as a
+        ON v.index = a.participant
+        WHERE
+            a.slot between `now` and `2 epoch ago` AND
+            a.justified_slot = previous_justified_slot AND
+            a.epoch_boundary_root = `2 epoch ago`
+        GROUP BY v.index
+
+    current_epoch_boundary_attesting_balance:
+        SELECT sum(v.balance)
+        FROM Validator as v
+        RIGHT JOIN Attestations as a
+        ON v.index = a.participant
+        WHERE
+            a.slot between `now` and `1 epoch ago` AND
+            a.justified_slot = justified_slot AND
+            a.epoch_boundary_root = `1 epoch ago`
+        GROUP BY v.index
+    """
+
+    EPOCH_LENGTH = config.EPOCH_LENGTH
+    MAX_DEPOSIT = config.MAX_DEPOSIT
+    LATEST_BLOCK_ROOTS_LENGTH = config.LATEST_BLOCK_ROOTS_LENGTH
+
+    current_epoch_attestations = tuple(
+        attestation
+        for attestation in state.latest_attestations
+        if state.slot - EPOCH_LENGTH <= attestation.data.slot < state.slot
+    )
+    previous_epoch_attestations = tuple(
+        attestation
+        for attestation in state.latest_attestations
+        if state.slot - 2 * EPOCH_LENGTH <= attestation.data.slot < state.slot - EPOCH_LENGTH
+    )
+
+    previous_epoch_justified_attestations = tuple(
+        attestation
+        for attestation in current_epoch_attestations + previous_epoch_attestations
+        if attestation.justified_slot == state.previous_justified_slot
+    )
+
+    previous_epoch_boundary_root = get_block_root(
+        state,
+        state.slot - 2 * EPOCH_LENGTH,
+        LATEST_BLOCK_ROOTS_LENGTH)
+    previous_epoch_boundary_attestations = tuple(
+        attestation
+        for attestation in previous_epoch_justified_attestations
+        if attestation.epoch_boundary_root == previous_epoch_boundary_root
+    )
+
+    sets_of_previous_epoch_boundary_participants = tuple(
+        frozenset(get_attestation_participants(
+            state,
+            attestation.data.slot,
+            attestation.data.shard,
+            attestation.participation_bitfield,
+            EPOCH_LENGTH,
+        ))
+        for attestation in previous_epoch_boundary_attestations
+    )
+    previous_epoch_boundary_attester_indices = (
+        tuple()
+        if len(sets_of_previous_epoch_boundary_participants) == 0
+        else frozenset.union(*sets_of_previous_epoch_boundary_participants)
```

Comment:
*if* it is possible for `set_of_previous_epoch_boundary_participants` to be empty then this line will fail.

Reply:
That's why I awkwardly protect it by checking `if len(sets_of_previous_epoch_boundary_participants) == 0`
This works on my Python:
```python
>>> sets_of_previous_epoch_boundary_participants = tuple()
>>> frozenset() if len(sets_of_previous_epoch_boundary_participants) == 0 else frozenset.union(*sets_of_previous_epoch_boundary_participants)
frozenset()
```
Is there a better way to do that union?

Reply:
:man_facepalming: sorry, previous comment was misguided but now I see why and thus the awkwardness.

I'd change this to be a `try/except`.  Removes the *checking* overhead and is more *pythonic* (depending on who you ask).

```python
try:
    previous_epoch_boundary_attester_indices = frozenset.union(*sets_of_previous_epoch_boundaries)
except:
    previous_epoch_boundary_attester_indices = frozenset()
```

Alternatively you could do this:

```python
previous_epoch_boundary_attester_indices = frozenset.union(frozenset(), *sets_of_previous_epoch_boundaries)
# or more simply
previous_epoch_boundary_attester_indices = frozenset().union(*sets_of_previous_epoch_boundaries)
```

Both of these seems to work correctly with an empty `sets_of_previous_epoch_boundaries`.

Reply:
A note for what I learned from the implementation.
I like this one for its expresiveness: 

```python
previous_epoch_boundary_attester_indices = frozenset().union(*sets_of_previous_epoch_boundaries)
```

but it's trickier than I've imagined.
Since the expression starts with an empty frozenset, mypy asked the type of the variable to be explicitly annotated. So it become this:

```python
previous_epoch_boundary_attester_indices: Iterable[ValidatorIndex]  = frozenset().union(*sets_of_previous_epoch_boundaries)
```
However, the variable annotation feature is not supported in Python3.5.

At the end of the day, this makes CI happy.

```python
    previous_epoch_boundary_attester_indices = frozenset.union(
        frozenset(),
        *sets_of_previous_epoch_boundary_participants,
    )
```